design/service-discovery.md

```diff
@@ -0,0 +1,178 @@
+
+### CustomClientInfo ###
+
+* Add a reference to CustomParticipantInfo, to be able to retrieve the relations between the incoming request and the subscriber that will be receiving the response.
+
+* Add `pending_matches_` to keep track of cases where the request publisher has matched with the server but the response subscriber has not. It will be an unordered map with the server response publisher's GUID as key:
+  `std::unordered_map<eprosima::fastrtps::rtps::GUID_t, eprosima::fastrtps::rtps::GUID_t, rmw_fastrtps_shared_cpp::hash_fastrtps_guid>`
+
+* Add `complete_matches_` to keep track of fully matched servers. It will be an unordered map with the server response publisher's GUID as key:
+  `std::unordered_map<eprosima::fastrtps::rtps::GUID_t, eprosima::fastrtps::rtps::GUID_t, rmw_fastrtps_shared_cpp::hash_fastrtps_guid>`
+
+* Add `complete_matches_count_` to hold the size of `complete_matches_`, to be used on `rmw_service_server_is_available`. It will be an atomic variable `std::atomic_size_t`.
+
+### ClientPubListener::OnPublicationMatched ###
+
+* When a request subscriber is found, if the response subscriber is matched against the server whose subscriber we just found, add the server's publisher/subscriber pair to the `complete_matches_` map. Else, add it to the `pending_matches_` map.
+* In the case of an unmatch, remove the pair for both `pending_matches_` and `complete_matches_`. Note that having only the response subscriber matched is not being tracked as pending match, as this will be resolved by the algorithm either when the response subscriber is unmatched or when the request publisher is matched again.
+* All these modifications only have to be applied when the relation between server's GUIDs are present in `remote_response_guid_by_remote_request_guid_`. If the matched subscriber's GUID is not there, it means that the server does not implement this solution and legacy behavior has to be applied.
+
+![ClientPubListener::OnPublicationMatched](service-discovery-images/ClientPubListener-OnPublicationMatched.svg)
+
+### ClientListener::OnSubscriptionMatched ###
+
+* When a response publisher is found, if the remote GUID is found on `pending_matches_`, we are about to complete the discovery for the response topic. Remove the entry from `pending_matches_` and move it to `complete_matches_`.
+* In the case of an unmatch, if the remote GUID is found on `complete_matches_`, we are about to have a half-discovery for the response topic. Remove the entry from `complete_matches_` and move it to `pending_matches_`.
+* Note that having only the response subscriber matched is not being tracked as pending match, so:
+  * In the case of a match, if the remote GUID is not on `pending_matches_`, it means that either the request subscriber is still umatched or that the remote server does not implement this solution. In either case there is nothing to do.
```

Comment:
@IkerLuengo did you mean
```suggestion
  * In the case of a match, if the remote GUID is not on `pending_matches_`, it means that either the request publisher is still umatched or that the remote server does not implement this solution. In either case there is nothing to do.
```
considering we're looking at it from the client POV? Same below.